### DISCUSSION TITLE
API-Level Hook for JunctionRestrictionsManager

### DISCUSSION THREAD

**Discussion by Elesbaan70** [CitiesSkylinesMods/TMPE#1557] [Ideas]
I would like to give the Node Controller mods a better way of interacting with `JunctionRestrictionsManager`, so that the code can be cleaned up, revised, and enhanced without having to worry so much about breaking Node Controller.

The idea is to use .NET events to provide a hook at the API level, which Node Controller can use to alter the default and configurable states of the flags, without having to rely on Harmony-based patching. I would do the work in Node Controller and provide pull requests once the new API in TMPE is released to stable.

Here's a rough sketch of what I have in mind:

```cs
        public enum JunctionRestrictionFlags {
            AllowUTurn = 1 << 0,
            AllowNearTurnOnRed = 1 << 1,
            AllowFarTurnOnRed = 1 << 2,
            AllowForwardLaneChange = 1 << 3,
            AllowEnterWhenBlocked = 1 << 4,
            AllowPedestrianCrossing = 1 << 5,
        }

        public class JunctionRestrictionsHookData {

            public JunctionRestrictionFlags Mask { get; }

            public JunctionRestrictionFlags Result { get; set; }
        }

        public delegate void JunctionRestrictionsHook(JunctionRestrictionsHookData hookData);

        public interface IJunctionRestrictionsManager {

            event JunctionRestrictionsHook DefaultJunctionRestrictionsHook;

            event JunctionRestrictionsHook ConfigurableJunctionRestrictionsHook;
        }
```

and it would be used something like this:

```cs
        public class SomeClass {

            public SomeClass() {
                JunctionRestrictionsManager.ConfigurableJunctionRestrictionsHook += ConfigurableJunctionRestrictionsHook;
            }

            public void ConfigurableJunctionRestrictionsHook(JunctionRestrictionsHookData hookData) {
                if ((hookData.Mask & JunctionRestrictionFlags.AllowUTurn) != 0) {

                    // do the work of figuring out whether this is configurable

                    if (result)
                        hookData.Result |= JunctionRestrictionFlags.AllowUTurn;
                    else
                        hookData.Result &= ~JunctionRestrictionFlags.AllowUTurn;
                }
            }

        }
```

@krzychu124 @kianzarrin @MacSergey Any thoughts on this?

**Comment by kianzarrin** [comment]
I can update my patches. macsurgey has been busy lately but might come back if there is a mod breaking change.

**Reply by kianzarrin** [reply]
the only problem is that how can NC support both stable and TEsT version of TMPE at the same time?

**Reply by kianzarrin** [reply]
I can use reflection but i fear that would be slow.

**Reply by Elesbaan70** [reply]
> macsurgey has been busy lately

Since I've contributed to @MacSergrey's mods before, my hope was that if this were discussed in advance, I'd be able to make the change for him in a way he'd be comfortable using with minimal effort on his part.

> the only problem is that how can NC support both stable and TEsT version of TMPE at the same time?

You won't have to worry about this if it's rolled out in stages:

1. New API released to stable
2. Node Controller updated to use new API, and released to stable
3. At some undefined point in the future, changes are made that would have broken the old way, but it just won't matter anymore

**Reply by kianzarrin** [reply]
@krzychu124 can this go to stable together with my API changes?

**Reply by Elesbaan70** [reply]
Keep in mind that I'm just discussing this in advance to get feedback and hopefully do it right the first time. I won't be coding any of this until after I've finished the new persistence. If you have other stuff that doesn't depend on this, I wouldn't suggest waiting on it.

**Reply by Elesbaan70** [reply]
> I can update my patches

The objective is for the patches on `JunctionRestrictionsManager` to become completely unnecessary.

**Reply by Elesbaan70** [reply]
@kianzarrin How does RM Crossings know when the pedestrian crossing flag changes? Is it patching `JunctionRestrictionsManager`? We could also add a `FlagsChanged` event.

I can't find the source for RM Crossings....

**Reply by krzychu124** [reply]
> I can't find the source for RM Crossings....

@Elesbaan70 https://github.com/CitiesSkylinesMods/HideCrosswalks

**Reply by Elesbaan70** [reply]
Ahhh, since Kian was listed as the mod's creator, I was looking for it in his repositories. 🙂

**Reply by kianzarrin** [reply]
I put link to my mods repo in my mods page and also in discord

**Reply by Elesbaan70** [reply]
> I put link to my mods repo in my mods page and also in discord

The link is buried deep in the text. I should have searched for "GitHub". 🙂
